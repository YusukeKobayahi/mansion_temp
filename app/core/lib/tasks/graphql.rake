# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
