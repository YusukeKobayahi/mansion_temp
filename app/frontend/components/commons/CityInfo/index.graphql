fragment MsLibraryCitiesCityInfo on City {
  name
  prefecture {
    name
  }
  info {
    crime {
      crimes
      crimeRatioPercentage
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
      area {
        id
        name
      }
    }
    household {
      households
      ownedHouseholds
      ownedPercentage
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    age {
      averageAge
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    income {
      taxpayers
      totalIncome
      averageIncome
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    medical {
      hospitalClinic {
        hospitals
        clinics
        perPopulation
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      bed {
        beds
        perPopulation
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
    }
    birthrate {
      birthrate
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    care {
      careHomes
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    park {
      parks
      parkArea
      areaPerPerson
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
    school {
      middleProceed {
        schoolProceed
        nationalSchoolProceed
        privateSchoolProceed
        proceedRatioPercentage
        prefectureRank
        capitalRank
        deviation
        grade {
          id
          typeName
        }
      }
      kindergarten {
        kindergartens
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      primary {
        primarySchools
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      middle {
        middleSchools
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      high {
        highSchools
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
    }
    population {
      growth {
        populationGrowthPercentage
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      density {
        populationDensity
        prefectureRank
        areaRank
        deviation
        grade {
          id
          typeName
        }
      }
      underFifteen {
        underFifteenPopulation
        underFifteenPopulationPercentage
        capitalRank
        prefectureRank
        deviation
        grade {
          id
          typeName
        }
      }
    }
    waitingChild {
      waitingChilds
      waitingChildsPerUnderFourPopulation
      prefectureRank
      areaRank
      deviation
      grade {
        id
        typeName
      }
    }
  }
}
