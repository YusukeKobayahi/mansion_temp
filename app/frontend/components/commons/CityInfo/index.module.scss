@import "~/styles/variables";
@import "~/styles/utils";

.container {
  padding-top: 50px;

  @include screenMd {
    padding-top: 40px;
  }
}

.inner {
  width: 678px;
  margin: 50px auto 0;

  @include screenMd {
    width: auto;
    padding: 0 16px;
    margin: 0;
  }
}

.table {
  width: 100%;
  tr {
    border-bottom: 1px solid #d0d0d0;
    font-size: 16px;

    @include screenMd {
      font-size: 13px;
    }

    &.none {
      display: none;
    }
  }

  th {
    text-align: center;
    padding: 14px;
    background-color: #faf9f7;
    vertical-align: middle;
    white-space: nowrap;
    color: #646464;

    @include screenMd {
      padding: 5px;
    }

    span {
      font-size: 13px;

      @include screenMd {
        font-size: 11px;
      }
    }

    &:nth-child(1) {
      width: 35%;

      @include screenMd {
        width: 40%;
      }
    }

    &:nth-child(n + 2) {
      width: 24%;

      @include screenMd {
        width: 20%;
      }
    }
  }

  td {
    text-align: center;
    padding: 10px;
    line-height: 1.4;
    padding: 14px;
    color: #646464;

    @include screenMd {
      padding: 5px;
    }

    &:nth-child(1) {
      text-align: left;
    }
  }
}

.btn {
  display: block;
  margin: 20px auto 0;
  padding: 8px 20px;
  background-color: #f8f6f0;
  border: 2px solid $color-main;
  border-radius: 40px;
  font-size: 16px;
  color: $color-main;
  cursor: pointer;

  @include screenMd {
    padding: 10px 30px;
    font-size: 14px;
  }
}
