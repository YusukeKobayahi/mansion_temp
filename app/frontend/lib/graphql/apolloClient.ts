// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
