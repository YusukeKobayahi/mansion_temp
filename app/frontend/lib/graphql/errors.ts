export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
