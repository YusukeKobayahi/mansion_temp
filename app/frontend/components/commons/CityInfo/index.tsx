import styles from "~/components/commons/CityInfo/index.module.scss";
import { MsLibraryCitiesCityInfoFragment } from "~/graphql/generated";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import SidelineHeading from "~/components/commons/headings/SidelineHeading";
import { yenFormat } from "~/lib/utils";
import { useState } from "react";
import classnames from "classnames";

export interface CityInfoProps {
  data: MsLibraryCitiesCityInfoFragment;
  className?: string;
}

const CityInfo: React.FC<CityInfoProps> = ({
  data,
  className,
}: CityInfoProps) => {
  const [viewState, setviewState] = useState(false);
  const info = data.info;
  if (
    !info.income &&
    !info.age &&
    !info.household &&
    !info.population &&
    !info.crime &&
    !info.medical &&
    !info.care &&
    !info.birthrate &&
    !info.park &&
    !info.school &&
    !info.waitingChild
  )
    return null;
  const checkUndefined = (n: number | string): number | string => {
    return n === 0 ? 0 : n ? n : "-";
  };

  const { name, prefecture } = data;
  return (
    <section className={classnames(styles.container, className)} id="local">
      <h2>
        <BoxHeading>
          数字で見る
          {prefecture?.name}
          {name}
        </BoxHeading>
      </h2>
      <div className={styles.inner}>
        <h3>
          <SidelineHeading>基本情報</SidelineHeading>
        </h3>
        <table className={styles.table}>
          <thead>
            <tr>
              <th></th>
              <th>実数値</th>
              <th>偏差値</th>
              <th>
                順位
                <br className={"spOnly"} />
                <span>({info.crime?.area?.name}内)</span>
              </th>
            </tr>
          </thead>
          <tbody>
            {info.income && (
              <tr>
                <td>平均年収</td>
                <td>
                  {checkUndefined(
                    yenFormat(Number(info.income?.averageIncome))
                  )}
                </td>
                <td>{checkUndefined(Number(info.income?.deviation))}</td>
                <td>{checkUndefined(Number(info.income?.areaRank))}位</td>
              </tr>
            )}
            {info.age && (
              <tr>
                <td>平均年齢</td>
                <td>{checkUndefined(Number(info.age?.averageAge))}歳</td>
                <td>{checkUndefined(Number(info.age?.deviation))}</td>
                <td>{checkUndefined(Number(info.age?.areaRank))}位</td>
              </tr>
            )}
            {info.birthrate && (
              <tr>
                <td>特殊出生率</td>
                <td>{checkUndefined(Number(info.birthrate?.birthrate))}％</td>
                <td>{checkUndefined(Number(info.birthrate?.deviation))}</td>
                <td>{checkUndefined(Number(info.birthrate?.areaRank))}位</td>
              </tr>
            )}
            {info.household && (
              <tr>
                <td>持ち家比率</td>
                <td>
                  {checkUndefined(Number(info.household?.ownedPercentage))}％
                </td>
                <td>{checkUndefined(Number(info.household?.deviation))}</td>
                <td>{checkUndefined(Number(info.household?.areaRank))}位</td>
              </tr>
            )}
            {info.population?.density && (
              <tr>
                <td>人口密度</td>
                <td>
                  {checkUndefined(
                    Number(info.population?.density?.populationDensity)
                  )}
                  <br className={"spOnly"} />
                  人/km2
                </td>
                <td>
                  {checkUndefined(Number(info.population?.density?.deviation))}
                </td>
                <td>
                  {checkUndefined(Number(info.population?.density?.areaRank))}位
                </td>
              </tr>
            )}
            {info.population?.growth && (
              <tr className={!viewState ? styles.none : ""}>
                <td>人口増加率</td>
                <td>
                  {checkUndefined(
                    Number(info.population?.growth?.populationGrowthPercentage)
                  )}
                  ％
                </td>
                <td>
                  {checkUndefined(Number(info.population?.growth?.deviation))}
                </td>
                <td>
                  {checkUndefined(Number(info.population?.growth?.areaRank))}位
                </td>
              </tr>
            )}
            {info.crime && (
              <tr className={!viewState ? styles.none : ""}>
                <td>
                  犯罪率
                  <br className={"spOnly"} />
                  (犯罪件数/人口)
                </td>
                <td>
                  {checkUndefined(Number(info.crime?.crimeRatioPercentage))}%
                </td>
                <td>{checkUndefined(Number(info.crime?.deviation))}</td>
                <td>{checkUndefined(Number(info.crime?.areaRank))}位</td>
              </tr>
            )}
            {info.medical?.hospitalClinic && (
              <tr className={!viewState ? styles.none : ""}>
                <td>病院・診療所数</td>
                <td>
                  {Number(info.medical?.hospitalClinic?.clinics) +
                    Number(info.medical?.hospitalClinic?.hospitals)}
                  件
                </td>
                <td>
                  {checkUndefined(
                    Number(info.medical?.hospitalClinic?.deviation)
                  )}
                </td>
                <td>
                  {checkUndefined(
                    Number(info.medical?.hospitalClinic?.areaRank)
                  )}
                  位
                </td>
              </tr>
            )}
            {info.care && (
              <tr className={!viewState ? styles.none : ""}>
                <td>介護施設数</td>
                <td>{checkUndefined(Number(info.care?.careHomes))}件</td>
                <td>{checkUndefined(Number(info.care?.deviation))}</td>
                <td>{checkUndefined(Number(info.care?.areaRank))}位</td>
              </tr>
            )}
            {info.park && (
              <tr className={!viewState ? styles.none : ""}>
                <td>
                  公園面積
                  <br className={"spOnly"} />
                  (10㎡/1人)
                </td>
                <td>
                  {checkUndefined(Number(info.park?.areaPerPerson))}
                  <br className={"spOnly"} />
                  ㎡/人
                </td>
                <td>{checkUndefined(Number(info.park?.deviation))}</td>
                <td>{checkUndefined(Number(info.park?.areaRank))}位</td>
              </tr>
            )}
            {info.waitingChild && (
              <tr className={!viewState ? styles.none : ""}>
                <td>
                  待機児童数
                  <br className={"spOnly"} />
                  (1,000人あたり)
                </td>
                <td>
                  {checkUndefined(Number(info.waitingChild?.waitingChilds))}人
                </td>
                <td>{checkUndefined(Number(info.waitingChild?.deviation))}</td>
                <td>{checkUndefined(Number(info.waitingChild?.areaRank))}位</td>
              </tr>
            )}
            {info.school?.kindergarten && (
              <tr className={!viewState ? styles.none : ""}>
                <td>幼稚園施設数</td>
                <td>
                  {checkUndefined(
                    Number(info.school?.kindergarten?.kindergartens)
                  )}
                  校
                </td>
                <td>
                  {checkUndefined(Number(info.school?.kindergarten?.deviation))}
                </td>
                <td>
                  {checkUndefined(Number(info.school?.kindergarten?.areaRank))}
                  位
                </td>
              </tr>
            )}
            {info.school?.primary && (
              <tr className={!viewState ? styles.none : ""}>
                <td>小学校施設数</td>
                <td>
                  {checkUndefined(Number(info.school?.primary?.primarySchools))}
                  校
                </td>
                <td>
                  {checkUndefined(Number(info.school?.primary?.deviation))}
                </td>
                <td>
                  {checkUndefined(Number(info.school?.primary?.areaRank))}位
                </td>
              </tr>
            )}
            {info.school?.middle && (
              <tr className={!viewState ? styles.none : ""}>
                <td>中学校施設数</td>
                <td>
                  {checkUndefined(Number(info.school?.middle?.middleSchools))}校
                </td>
                <td>
                  {checkUndefined(Number(info.school?.middle?.deviation))}
                </td>
                <td>
                  {checkUndefined(Number(info.school?.middle?.areaRank))}位
                </td>
              </tr>
            )}
            {info.school?.middleProceed && (
              <tr className={!viewState ? styles.none : ""}>
                <td>国私立中学校進学率</td>
                <td>
                  {checkUndefined(
                    Number(info.school?.middleProceed?.proceedRatioPercentage)
                  )}
                  ％
                </td>
                <td>
                  {checkUndefined(
                    Number(info.school?.middleProceed?.deviation)
                  )}
                </td>
                <td>
                  {checkUndefined(
                    Number(info.school?.middleProceed?.capitalRank)
                  )}
                  位
                </td>
              </tr>
            )}
          </tbody>
        </table>
        {!viewState && (
          <button onClick={() => setviewState(true)} className={styles.btn}>
            もっと見る
          </button>
        )}
      </div>
    </section>
  );
};

export default CityInfo;
