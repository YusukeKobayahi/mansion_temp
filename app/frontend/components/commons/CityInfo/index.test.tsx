import { render } from "@testing-library/react";
import mockData from "~/components/pages/mansions/fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import CityInfo from "~/components/commons/CityInfo";

describe("components/commons/CityInfo", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<CityInfo data={mock.mansion.city} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
