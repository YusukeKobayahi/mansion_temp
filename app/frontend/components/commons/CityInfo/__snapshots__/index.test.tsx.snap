// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/commons/CityInfo 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
    id="local"
  >
    <h2>
      <div
        class="heading"
      >
        数字で見る千代田区
      </div>
    </h2>
    <div
      class="inner"
    >
      <h3>
        <div
          class="heading"
        >
          <span
            class="inner"
          >
            基本情報
          </span>
        </div>
      </h3>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th />
            <th>
              実数値
            </th>
            <th>
              偏差値
            </th>
            <th>
              順位
              <br
                class="spOnly"
              />
              <span>
                (一都三県内)
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              平均年収
            </td>
            <td>
              555万5,123円
            </td>
            <td>
              69.88
            </td>
            <td>
              8位
            </td>
          </tr>
          <tr>
            <td>
              平均年齢
            </td>
            <td>
              44歳
            </td>
            <td>
              41.49
            </td>
            <td>
              213位
            </td>
          </tr>
          <tr>
            <td>
              特殊出生率
            </td>
            <td>
              0.99％
            </td>
            <td>
              23.08
            </td>
            <td>
              248位
            </td>
          </tr>
          <tr>
            <td>
              持ち家比率
            </td>
            <td>
              32.8％
            </td>
            <td>
              27.85
            </td>
            <td>
              210位
            </td>
          </tr>
          <tr>
            <td>
              人口密度
            </td>
            <td>
              18307.4
              <br
                class="spOnly"
              />
              人/km2
            </td>
            <td>
              75.11
            </td>
            <td>
              8位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              人口増加率
            </td>
            <td>
              0.99％
            </td>
            <td>
              60.16
            </td>
            <td>
              30位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              犯罪率
              <br
                class="spOnly"
              />
              (犯罪件数/人口)
            </td>
            <td>
              1.36%
            </td>
            <td>
              84.14
            </td>
            <td>
              3位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              病院・診療所数
            </td>
            <td>
              659件
            </td>
            <td>
              65.4326
            </td>
            <td>
              8位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              介護施設数
            </td>
            <td>
              50件
            </td>
            <td>
              52.21
            </td>
            <td>
              77位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              公園面積
              <br
                class="spOnly"
              />
              (10㎡/1人)
            </td>
            <td>
              3.3
              <br
                class="spOnly"
              />
              ㎡/人
            </td>
            <td>
              47.78
            </td>
            <td>
              184位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              待機児童数
              <br
                class="spOnly"
              />
              (1,000人あたり)
            </td>
            <td>
              0人
            </td>
            <td>
              45.95
            </td>
            <td>
              126位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              幼稚園施設数
            </td>
            <td>
              30校
            </td>
            <td>
              68.21
            </td>
            <td>
              18位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              小学校施設数
            </td>
            <td>
              30校
            </td>
            <td>
              60.82
            </td>
            <td>
              25位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              中学校施設数
            </td>
            <td>
              16校
            </td>
            <td>
              59.34
            </td>
            <td>
              27位
            </td>
          </tr>
          <tr
            class="none"
          >
            <td>
              国私立中学校進学率
            </td>
            <td>
              34.34％
            </td>
            <td>
              64.97
            </td>
            <td>
              7位
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="btn"
      >
        もっと見る
      </button>
    </div>
  </section>
</DocumentFragment>
`;
